src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import copy
+import os
+from dataclasses import dataclass
+from typing import List
+
+import yaml
+from jinja2 import Environment, UndefinedError, nodes
+from snowflake.cli.api.project.schemas.project_definition import ProjectDefinition
+from snowflake.cli.api.utils.rendering import get_snowflake_cli_jinja_env
+
+
+@dataclass
+class GraphNode:
+    path: str
+    value: str | None = None
+
+
+def _get_referenced_vars(ast_node, attr_chain: List[str] = []) -> List[List[str]]:
+    # Var nodes end in Name node, and starts with Getattr node.
+    # Example: ctx.env.test will look like Getattr(test) -> Getattr(env) -> Name(ctx)
+    all_referenced_vars = []
+
+    if isinstance(ast_node, nodes.Getattr):
+        attr_chain = [getattr(ast_node, "attr")] + attr_chain
+    elif isinstance(ast_node, nodes.Name):
+        all_referenced_vars = [[getattr(ast_node, "name")] + attr_chain]
+
+    for child_node in ast_node.iter_child_nodes():
+        all_referenced_vars = all_referenced_vars + _get_referenced_vars(
+            child_node, attr_chain
+        )
+
+    return all_referenced_vars
+
+
+def _get_referenced_vars_from_str(env: Environment, template_str: str) -> List[str]:
+    ast = env.parse(template_str)
+    referenced_vars = _get_referenced_vars(ast)
+    result = [".".join(vars_parts) for vars_parts in referenced_vars]
+    return result
+
+
+def _get_value_from_var_path(env: Environment, context: dict, var_path: str):
+    # given a variable path (e.g. ctx.env.test), return evaluated value based on context
+    # fall back to env variables and escape them so we stop the chain of variables
+    ref_str = "<% " + var_path + " %>"
+    template = env.from_string(ref_str)
+    try:
+        return template.render(context)
+    except UndefinedError as e:
+        # check in env variables
+        try:
+            return f"{env.block_start_string} raw {env.block_end_string}{template.render({'ctx': {'env': os.environ}})}{env.block_start_string} endraw {env.block_end_string}"
```

Comment:
I don't know what's happening there

Reply:
when evaluating environment variables coming from os, we do not want to expand them further, which is why i am using {% raw %} {% endraw %} templates. I don't like that though. I will see if I can do better.